cmd/nerdctl/compose_exec_linux_test.go

```diff
@@ -0,0 +1,135 @@
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/containerd/nerdctl/pkg/testutil"
+)
+
+func TestComposeExec(t *testing.T) {
+	// `-i` in `compose run & exec` is only supported in compose v2.
+	// Currently CI is using compose v1.
+	testutil.DockerIncompatible(t)
+	base := testutil.NewBase(t)
+	var dockerComposeYAML = fmt.Sprintf(`
+version: '3.1'
+
+services:
+  svc0:
+    image: %s
+  svc1:
+    image: %s
+`, testutil.CommonImage, testutil.CommonImage)
+
+	comp := testutil.NewComposeDir(t, dockerComposeYAML)
+	defer comp.CleanUp()
+	projectName := comp.ProjectName()
+	t.Logf("projectName=%q", projectName)
+
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "run", "-d", "-i=false", "svc0", "sleep", "1h").AssertOK()
+	defer base.ComposeCmd("-f", comp.YAMLFullPath(), "down", "-v").AssertOK()
+
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "exec", "-i=false", "-t=false", "svc0", "echo", "success").AssertOutExactly("success\n")
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "exec", "-i=false", "-t=false", "--workdir", "/tmp", "svc0", "pwd").AssertOutExactly("/tmp\n")
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "exec", "-i=false", "-t=false", "svc1", "echo", "success").AssertFail()
+}
+
+func TestComposeExecWithUser(t *testing.T) {
+	// `-i` in `compose run & exec` is only supported in compose v2.
+	// Currently CI is using compose v1.
+	testutil.DockerIncompatible(t)
+	base := testutil.NewBase(t)
+	var dockerComposeYAML = fmt.Sprintf(`
+version: '3.1'
+
+services:
+  svc0:
+    image: %s
+  svc1:
+    image: %s
+`, testutil.CommonImage, testutil.CommonImage)
+
+	comp := testutil.NewComposeDir(t, dockerComposeYAML)
+	defer comp.CleanUp()
+	projectName := comp.ProjectName()
+	t.Logf("projectName=%q", projectName)
+
+	testContainer := testutil.Identifier(t)
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "run", "-d", "-i=false", "--name", testContainer, "svc0", "sleep", "1h").AssertOK()
+	defer base.ComposeCmd("-f", comp.YAMLFullPath(), "down", "-v").AssertOK()
+	base.EnsureContainerStarted(testContainer)
+
+	testCases := map[string]string{
+		"":             "uid=0(root) gid=0(root)",
+		"1000":         "uid=1000 gid=0(root)",
+		"1000:users":   "uid=1000 gid=100(users)",
+		"guest":        "uid=405(guest) gid=100(users)",
+		"nobody":       "uid=65534(nobody) gid=65534(nobody)",
+		"nobody:users": "uid=65534(nobody) gid=100(users)",
+	}
+
+	for userStr, expected := range testCases {
+		args := []string{"-f", comp.YAMLFullPath(), "exec", "-i=false", "-t=false"}
+		if userStr != "" {
+			args = append(args, "--user", userStr)
+		}
+		args = append(args, "svc0", "id")
+		base.ComposeCmd(args...).AssertOutContains(expected)
```

Comment:
AssertOutExactly might be preferable

Reply:
Thanks for the suggestion.

`id` output (the empty case) has extra strings:

```
➜  compose-exec sndc exec svc0 id
uid=0(root) gid=0(root) groups=1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),20(dialout),26(tape),27(video)
➜  compose-exec sndc exec -u 1000 svc0 id
uid=1000 gid=0(root)
```